src/atomic/atomic.lisp

```diff
@@ -10,21 +10,29 @@
   cell)
 
 (defun make-atomic-reference (&key (value nil))
-  (%make-atomic-reference value))
+  #+ccl (%make-atomic-reference (make-array 1 :initial-element value))
+  #-ccl (%make-atomic-reference (cons value nil)))
 
 (defmethod print-object ((ref atomic-reference) stream)
   (print-unreadable-object (ref stream :type t :identity t)
     (format stream "~S" (atomic-get ref))))
 
+(defmacro atomic-place (ref)
+  #+ccl
+  `(svref (slot-value ,ref 'cell) 0)
+  #-ccl
+  `(car (slot-value ,ref 'cell)))
+
 (defmethod atomic-cas ((ref atomic-reference) expected new)
-  (let* ((*the-ref* (slot-value ref 'cell))
-         (cas-result (atomics:cas *the-ref* expected new)))
-    (when cas-result
-      (setf (slot-value ref 'cell) *the-ref*))
-    cas-result))
+  (atomics:cas (atomic-place ref) expected new))
 
 (defmethod atomic-get ((ref atomic-reference))
-  (slot-value ref 'cell))
+  (atomic-place ref))
+
+(defmethod atomic-setf ((ref atomic-reference) new)
```

Comment:
So I would actually think that the loop should be in `atomic-cas`. `atomic-setf` is slightly higher-level and should provide the 'old' value to `atomic-cas` as parameter.